### PR TITLE
azurerm_cosmosdb_account add minimal_tls_version

### LINKED ISSUE #21295: Support for minimalTlsVersion in azurerm_cosmosdb_account
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I would like to request a new feature for the azurerm_cosmosdb_account in the Azure Resource Manager API. With the release of the 2022-11-15 API version of the Azure Cosmos DB Resource Provider API, it is now possible to set the minimum TLS version for Cosmos DB accounts.

Currently, the only way to set the minimum TLS version is through the [Azure API](https://learn.microsoft.com/en-us/azure/cosmos-db/self-serve-minimum-tls-enforcement). However, it would be beneficial for users to have the ability to set the minimum TLS value to 1.2 through the [azurerm_cosmosdb_account ](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/cosmosdb_account) resource.

Although it is a standard for new Cosmos DB accounts to have a minimum TLS version of 1.2 since April 1, 2023, this feature would still be useful for users who have existing Cosmos DB accounts and need to update the minimum TLS version.
Possible values for `minimalTlsVersion `could be:

- `Tls` for setting the minimum version to TLS 1.0.
- `Tls11` for setting the minimum version to TLS 1.1.
- `Tls12` for setting the minimum version to TLS 1.1.

I believe that adding this feature to `azurerm_cosmosdb_account `would greatly improve the user experience for managing Cosmos DB accounts through the Azure Resource Manager API.

Thank you for considering my feature request.

### New or Affected Resource(s)/Data Source(s)

azurerm_cosmosdb_account

### Potential Terraform Configuration

```hcl
resource "azurerm_cosmosdb_account" "db" {
  name                = "tfex-cosmos-db-${random_integer.ri.result}"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  offer_type          = "Standard"
  kind                = "MongoDB"

  minimalTlsVersion   = "Tls12"
}
```

### References

Azure TLS version enforcement documentation: https://learn.microsoft.com/en-us/azure/cosmos-db/self-serve-minimum-tls-enforcement

### PR DESCRIPTION
**Pull request by rizkybiz**
## Test Output
```
make acctests SERVICE='cosmos' TESTARGS='-run=TestAccCosmosDBAccount_MinimalTlsVersion' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/cosmos -run=TestAccCosmosDBAccount_MinimalTlsVersion -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccCosmosDBAccount_MinimalTlsVersion
=== PAUSE TestAccCosmosDBAccount_MinimalTlsVersion
=== CONT  TestAccCosmosDBAccount_MinimalTlsVersion
--- PASS: TestAccCosmosDBAccount_MinimalTlsVersion (723.97s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/cosmos        730.340s
```

fixes #21295

### CODE REVIEW

**Comment by neil-yechenwei** [comment]
Service team confirmed that there is a known issue that the default value of `minimalTlsVersion` of the source CosmosDB Account returned by service API is `Tls12` when it isn't set in the request payload but the default value of `minimalTlsVersion` of the restored CosmosDB Account returned by service API is `Tls` when it isn't set in the request payload. So service team is still investigating this issue.

Note: TestAccCosmosDBAccount_restoreCreateMode would fail once this PR is applied.

**Comment by jackofallops** [comment]
HI @rizkybiz - Thanks for this PR. Unfortunately, as noted in an earlier PR here https://github.com/hashicorp/terraform-provider-azurerm/pull/24615#discussion_r1469508193  this property has an upstream API issue which we are working with MSFT to resolve.  Due to this  ongoing work with the service team, I'm going to close this for now and we can take a look again when the upstream issue is resolved.
Thanks again!

File: internal/services/cosmos/cosmosdb_account_resource.go [hunk 1]
```diff
@@ -301,6 +301,12 @@ func resourceCosmosDbAccount() *pluginsdk.Resource {
 				}
 			}(),
 
+			"minimal_tls_version": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
```
Comment (katbyte) [inline internal/services/cosmos/cosmosdb_account_resource.go:306]: should we be specifying a default value here? or making it computed?

File: internal/services/cosmos/cosmosdb_account_resource.go [hunk 2]
```diff
@@ -746,6 +752,16 @@ func resourceCosmosDbAccountCreate(d *pluginsdk.ResourceData, meta interface{})
...
 
+	var minimalTlsVersion cosmosdb.MinimalTlsVersion
+	switch v := d.Get("minimal_tls_version").(string); v {
+	case "Tls":
+		minimalTlsVersion = cosmosdb.MinimalTlsVersionTls
+	case "Tls11":
+		minimalTlsVersion = cosmosdb.MinimalTlsVersionTlsOneOne
+	case "Tls12":
+		minimalTlsVersion = cosmosdb.MinimalTlsVersionTlsOneTwo
+	}
```
Comment (katbyte) [inline internal/services/cosmos/cosmosdb_account_resource.go:763]: how come we need this? shouldn't the allowed values match the constant?

**Review by katbyte — CHANGES_REQUESTED** [review]
Hey @rizkybiz - While there is an underlaying API bug this is something i think we can work around by in the create function once we’ve restored the CosmosDB account we can do an update on the other values (e.g. tls) to ensure they match the configuration

**Pull request closed without merging**